tools/lsp/language.rs

```diff
@@ -97,24 +98,170 @@ pub fn request_state(ctx: &std::rc::Rc<Context>) {
+            "Button" | "CheckBox" | "ComboBox" | "Slider" | "SpinBox" | "Switch" => "Input",
+            "ProgressIndicator" | "Spinner" => "Status",
+            "ListView" | "StandardListView" | "StandardTableView" => "Views",
+            _ => "Widgets",
+        }
+    };
+
+    ComponentInformation {
+        name: name.to_string(),
+        category: category.to_string(),
+        is_builtin: false,
+        is_std_widget: true,
+        is_local: false,
+        file: None,
+        offset: 0,
+    }
+}
+
+fn exported_project_component_info(name: &str, file: String, offset: u32) -> ComponentInformation {
+    ComponentInformation {
+        name: name.to_string(),
+        category: "User Defined".to_string(),
+        is_builtin: false,
+        is_std_widget: false,
+        is_local: false,
+        file: Some(file),
+        offset,
+    }
+}
+
+fn file_local_component_info(name: &str, file: String, offset: u32) -> ComponentInformation {
+    ComponentInformation {
+        name: name.to_string(),
+        category: "User Defined".to_string(),
+        is_builtin: false,
+        is_std_widget: false,
+        is_local: false,
+        file: Some(file),
+        offset,
+    }
+}
+
 /// A cache of loaded documents
 pub struct DocumentCache {
     pub(crate) documents: TypeLoader,
     preview_config: PreviewConfig,
+    builtin_elements: OnceCell<Vec<ComponentInformation>>,
 }
 
 impl DocumentCache {
     pub fn new(config: CompilerConfiguration) -> Self {
         let documents =
             TypeLoader::new(TypeRegister::builtin(), config, &mut BuildDiagnostics::default());
-        Self { documents, preview_config: Default::default() }
+        Self { documents, preview_config: Default::default(), builtin_elements: Default::default() }
     }
 
     pub fn document_version(&self, target_uri: &lsp_types::Url) -> SourceFileVersion {
         self.documents
             .get_document(&uri_to_file(target_uri).unwrap_or_default())
             .and_then(|doc| doc.node.as_ref()?.source_file.version())
     }
+
+    pub fn builtin_components(&self) -> impl Iterator<Item = &ComponentInformation> {
+        self.builtin_elements
+            .get_or_init(|| {
+                let registry = self.documents.global_type_registry.borrow();
+                let builtins = {
+                    let mut tmp: Vec<_> = registry
+                        .all_elements()
+                        .iter()
+                        .filter_map(|(name, ty)| {
+                            matches!(ty, ElementType::Builtin(b) if !b.is_internal)
+                                .then_some(builtin_component_info(name))
+                        })
+                        .collect();
+                    tmp.sort_by(|e1, e2| e1.name.cmp(&e2.name));
+                    tmp
+                };
+                builtins
+            })
+            .iter()
+    }
+
+    pub fn all_exported_components(&self, result: &mut Vec<ComponentInformation>) {
```

Comment:
I also wouldn't put this as a associated method of the DocumentCache, but rather a free function in the collection  module